File: src/TestExaminerViewer.cpp
/*
 * gleem -- OpenGL Extremely Easy-To-Use Manipulators.
 * Copyright (C) 1998 Kenneth B. Russell (kbrussel@media.mit.edu)
 * See the file LICENSE.txt in the doc/ directory for licensing terms.
 */

#ifdef WIN32
# include <windows.h>
#endif
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>
#include "../include/ManipManager.h"
#include "../include/HandleBoxManip.h"
#include "../include/MathUtil.h"
#include "../include/ExaminerViewer.h"

GLEEM_USE_NAMESPACE

HandleBoxManip *manip = NULL;
ExaminerViewer *viewer = NULL;

class HandleBoxManipBSphereProvider : 
  public ExaminerViewer::BSphereProvider
{
public:
  HandleBoxManipBSphereProvider(HandleBoxManip *manip)
  {
    this->manip = manip;
  }

  virtual const BSphere &getBoundingSphere()
  {
    bsph.setCenter(manip->getTranslation());
    GleemV3f scale0 = manip->getScale();
    GleemV3f scale1 = manip->getGeometryScale();
    GleemV3f scale;
    scale[0] = 2.0f * scale0[0] * scale1[0];
    scale[1] = 2.0f * scale0[1] * scale1[1];
    scale[2] = 2.0f * scale0[2] * scale1[2];
    bsph.setRadius(sqrtf(scale[0] * scale[0] +
			 scale[1] * scale[1] +
			 scale[2] * scale[2]));
    return bsph;
  }

private:
  HandleBoxManip *manip;
  BSphere bsph;
};

void
display()
{
  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
  viewer->update();
  ManipManager::getManipManager()->
    updateCameraParameters(viewer->getWindowID(),
			   viewer->getCameraParameters());
  ManipManager::getManipManager()->render();
  glFlush();
  glutSwapBuffers();
  glutPostRedisplay();
}

int
main(int argc, char **argv)
{
  glutInit(&argc, argv);

  glutInitDisplayMode(GLUT_RGB | GLUT_DOUBLE | GLUT_DEPTH);

  ManipManager::init(false);
  viewer = new ExaminerViewer();
  viewer->makeCurrent();
  viewer->setMouseDelegate(ManipManager::mouseFunc);
  viewer->setMotionDelegate(ManipManager::motionFunc);
  glutPassiveMotionFunc(ManipManager::passiveMotionFunc);

  //GLfloat matSpecular[] = {1, 1, 1, 1};
  //  GLfloat matShininess[] = {50};
  GLfloat lightPosition[] = {1, 1, 1, 0};
  glClearColor(0, 0, 0, 0);
  //  glShadeModel(GL_SMOOTH);
  //  glMaterialfv(GL_FRONT, GL_SPECULAR, matSpecular);
  //  glMaterialfv(GL_FRONT, GL_SHININESS, matShininess);
  GLfloat ambient[] = { 0.0, 0.0, 0.0, 1.0 };
  GLfloat diffuse[] = { 1.0, 1.0, 1.0, 1.0 };
  glLightfv(GL_LIGHT0, GL_AMBIENT, ambient);
  glLightfv(GL_LIGHT0, GL_DIFFUSE, diffuse);
  glLightfv(GL_LIGHT0, GL_POSITION, lightPosition);

  glEnable(GL_LIGHTING);
  glEnable(GL_LIGHT0);
  glEnable(GL_DEPTH_TEST);

  glutDisplayFunc(display);
  manip = new HandleBoxManip();
  //  manip->setGeometryScale(GleemV3f(1, 2, 1));
  //  manip->setTranslation(GleemV3f(0, 0, -10));
  viewer->setBSphereProvider(new HandleBoxManipBSphereProvider(manip));
  viewer->viewAll();

  glutMainLoop();
  return 0;
}

File: gl4java-impl/TestExaminerViewer.java
/*
 * gleem -- OpenGL Extremely Easy-To-Use Manipulators.
 * Copyright (C) 1998, 1999, 2002 Kenneth B. Russell (kbrussel@alum.mit.edu)
 * See the file LICENSE.txt in the doc/ directory for licensing terms.
 */

package gnu.gleem;

import java.awt.*;
import java.awt.event.*;
import gl4java.*;
import gl4java.awt.*;
import gl4java.drawable.*;
import gnu.gleem.linalg.*;

/** Tests the Examiner Viewer. */

public class TestExaminerViewer {
  private static final int X_SIZE = 400;
  private static final int Y_SIZE = 400;

  static class HandleBoxManipBSphereProvider implements BSphereProvider {
    private HandleBoxManip manip;

    private HandleBoxManipBSphereProvider(HandleBoxManip manip) {
      this.manip = manip;
    }

    public BSphere getBoundingSphere() {
      BSphere bsph = new BSphere();
      bsph.setCenter(manip.getTranslation());
      Vec3f scale0 = manip.getScale();
      Vec3f scale1 = manip.getGeometryScale();
      Vec3f scale = new Vec3f();
      scale.setX(2.0f * scale0.x() * scale1.x());
      scale.setY(2.0f * scale0.y() * scale1.y());
      scale.setZ(2.0f * scale0.z() * scale1.z());
      bsph.setRadius(scale.length());
      return bsph;
    }
  }

  static class Listener implements GLEventListener, GLEnum {
    private GLFunc gl;
    private GLUFunc glu;
    private CameraParameters params = new CameraParameters();
    private ExaminerViewer viewer;

    public void init(GLDrawable drawable) {
      gl = drawable.getGL();
      glu = drawable.getGLU();

      gl.glClearColor(0, 0, 0, 0);
      float[] lightPosition = new float[] {1, 1, 1, 0};
      float[] ambient       = new float[] { 0.0f, 0.0f, 0.0f, 1.0f };
      float[] diffuse       = new float[] { 1.0f, 1.0f, 1.0f, 1.0f };
      gl.glLightfv(GL_LIGHT0, GL_AMBIENT,  ambient);
      gl.glLightfv(GL_LIGHT0, GL_DIFFUSE,  diffuse);
      gl.glLightfv(GL_LIGHT0, GL_POSITION, lightPosition);

      gl.glEnable(GL_LIGHTING);
      gl.glEnable(GL_LIGHT0);
      gl.glEnable(GL_DEPTH_TEST);

      params.setPosition(new Vec3f(0, 0, 0));
      params.setForwardDirection(Vec3f.NEG_Z_AXIS);
      params.setUpDirection(Vec3f.Y_AXIS);
      params.setVertFOV((float) (Math.PI / 8.0));
      params.setImagePlaneAspectRatio(1);
      params.xSize = X_SIZE;
      params.ySize = Y_SIZE;

      gl.glMatrixMode(GL_PROJECTION);
      gl.glLoadIdentity();
      glu.gluPerspective(45, 1, 1, 100);
      gl.glMatrixMode(GL_MODELVIEW);
      gl.glLoadIdentity();

      // Register the window with the ManipManager
      ManipManager manager = ManipManager.getManipManager();
      manager.registerWindow(drawable);

      // Instantiate a HandleBoxManip
      HandleBoxManip manip = new HandleBoxManip();
      manip.setTranslation(new Vec3f(0, 0, -10));
      manager.showManipInWindow(manip, drawable);

      // Instantiate ExaminerViewer
      viewer = new ExaminerViewer(MouseButtonHelper.numMouseButtons());
      viewer.attach(drawable, new HandleBoxManipBSphereProvider(manip));
      viewer.viewAll(gl, glu);
    }

    public void display(GLDrawable drawable) {
      gl.glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
      viewer.update(gl, glu);
      ManipManager.getManipManager().updateCameraParameters(drawable, viewer.getCameraParameters());
      ManipManager.getManipManager().render(drawable, gl);
    }

    // Unused routines
    public void reshape(GLDrawable drawable, int w, int h) {}
    public void preDisplay(GLDrawable drawable) {}
    public void postDisplay(GLDrawable drawable) {}
    public void cleanup(GLDrawable drawable) {}
  }

  public static void main(String[] args) {
    Frame frame = new Frame("ExaminerViewer Test");
    frame.addWindowListener(new WindowAdapter() {
        public void windowClosing(WindowEvent e) {
          System.exit(0);
        }
      });
    frame.setLayout(new BorderLayout());
    GLCanvas canvas =
      GLDrawableFactory.getFactory().createGLCanvas(new GLCapabilities(), 400, 400);
    canvas.addGLEventListener(new Listener());
    frame.add(canvas, BorderLayout.CENTER);
    frame.pack();
    frame.show();
  }
}
